plugins/obs-ffmpeg/texture-amf.cpp

```diff
@@ -1384,10 +1437,77 @@ try {
 	return false;
 }
 
+static inline void check_recommended_avc_defaults(amf_base *enc,
+						  obs_data_t *settings)
+{
+	int64_t framerate = enc->fps_num / enc->fps_den;
+	if ((enc->cx * enc->cy > 1920 * 1088) || (framerate > 60)) {
+		// Recommended base defaults
+		obs_data_set_default_int(settings, "bitrate", 2500);
+		obs_data_set_default_int(settings, "cqp", 20);
+		obs_data_set_default_string(settings, "rate_control", "CBR");
+		obs_data_set_default_string(settings, "preset", "quality");
+		obs_data_set_default_string(settings, "profile", "high");
+		obs_data_set_default_int(settings, "bf", 0);
+		info("Original default settings were lowered according to resolution"
+		     " and framerate.");
+	} else {
+		/* Recommended high perceptual quality defaults
+		 * for up to 1080p60fps
+		 */
+
+		bool adaptive_minigop_enabled = true;
+		set_avc_property(enc, ADAPTIVE_MINIGOP,
+				 adaptive_minigop_enabled);
+
+		bool pa_enabled = obs_data_get_bool(settings, "pre_analysis");
+		bool pa_set_by_user =
+			obs_data_has_user_value(settings, "pre_analysis");
+
+		if (!pa_set_by_user || (pa_set_by_user && pa_enabled)) {
+			pa_enabled = true;
+			uint64_t pa_lab_depth = 20;
+			AMF_MEMORY_TYPE pa_engine_type = AMF_MEMORY_OPENCL;
+			AMF_PA_TAQ_MODE_ENUM pa_taq_mode = AMF_PA_TAQ_MODE_2;
+			obs_data_set_bool(settings, "pre_analysis", pa_enabled);
+
+			set_avc_property(enc, PRE_ANALYSIS_ENABLE, pa_enabled);
+			set_amf_property(enc, AMF_PA_LOOKAHEAD_BUFFER_DEPTH,
+					 pa_lab_depth);
+			set_amf_property(enc, AMF_PA_TAQ_MODE, pa_taq_mode);
+			set_amf_property(enc, AMF_PA_ENGINE_TYPE,
+					 AMF_MEMORY_OPENCL);
+
+			info("High perceptual quality defaults:\n"
+			     "\tAdaptiveMiniGOP:        %s\n"
+			     "\tEnablePreAnalysis:      %s\n"
+			     "\tPALookAheadBufferDepth: %d\n"
+			     "\tPATemporalAQMode:       %d\n"
+			     "\tPAEngineType:           %s\n",
+			     adaptive_minigop_enabled ? "true" : "false",
+			     pa_enabled ? "true" : "false", pa_lab_depth,
+			     pa_taq_mode,
+			     convert_amf_memory_type_to_string(pa_engine_type));
+		} else {
+			info("High perceptual quality defaults:\n"
+			     "\tAdaptiveMiniGOP:        %s\n",
+			     adaptive_minigop_enabled ? "true" : "false");
+		}
+	}
+}
+
 static bool amf_avc_init(void *data, obs_data_t *settings)
 {
 	amf_base *enc = (amf_base *)data;
 
+	/* We originally set the recommended defaults for high perceptual quality in
+	 * the UI settings. Once the resolution and framerate are available here
+	 * during init, we check if the defaults need to be lowered to the base
+	 * defaults or the remaining recommended high perceptual quality settings can
+	 * be set.
+	 */
```

Comment:
I dislike that we repeat this large block comment three times, but I'm not sure how to address that other than making it less verbose.

Reply:
Fixed in [409550a](https://github.com/obsproject/obs-studio/pull/9352/commits/409550a6bc3a9b1250ca0a538b90e23cd6757f20)